src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs

```diff
@@ -37,6 +37,43 @@ public sealed class AnalyzerConfigSet
         private readonly ConcurrentDictionary<ReadOnlyMemory<char>, string> _diagnosticIdCache =
             new ConcurrentDictionary<ReadOnlyMemory<char>, string>(CharMemoryEqualityComparer.Instance);
 
+        // PERF: Most files will probably have the same options, so share the dictionary instances
+        private readonly ConcurrentCache<List<Section>, AnalyzerConfigOptionsResult> _optionsCache =
+            new ConcurrentCache<List<Section>, AnalyzerConfigOptionsResult>(50, SequenceEqualComparer.Instance); // arbitrary size
+
+        private readonly ObjectPool<TreeOptions.Builder> _treeOptionsPool =
+            new ObjectPool<TreeOptions.Builder>(() => ImmutableDictionary.CreateBuilder<string, ReportDiagnostic>(Section.PropertiesKeyComparer));
+
+        private readonly ObjectPool<AnalyzerOptions.Builder> _analyzerOptionsPool =
+            new ObjectPool<AnalyzerOptions.Builder>(() => ImmutableDictionary.CreateBuilder<string, string>(Section.PropertiesKeyComparer));
+
+        private readonly ObjectPool<List<Section>> _sectionKeyPool = new ObjectPool<List<Section>>(() => new List<Section>());
+
+        private sealed class SequenceEqualComparer : IEqualityComparer<List<Section>>
+        {
+            public static SequenceEqualComparer Instance { get; } = new SequenceEqualComparer();
+
+            public bool Equals(List<Section> x, List<Section> y)
+            {
+                if (x.Count != y.Count)
+                {
+                    return false;
+                }
+
+                for (int i = 0; i < x.Count; i++)
+                {
+                    if (!ReferenceEquals(x[i], y[i]))
+                    {
+                        return false;
+                    }
+                }
+
+                return true;
```

Comment:
Can this be implemented with `System.Linq.Enumerable.SequenceEqual()`?
```
return x.SequenceEqual(y);
```

Reply:
No constant-time easy-out for incorrect lengths.